proto/asn1-pdu/x509_certificate_to_der.cc

```diff
@@ -43,6 +43,115 @@ DECLARE_ENCODE_FUNCTION(AlgorithmIdentifierSequence) {
   EncodeTagAndLength(kAsn1Sequence, der.size() - tag_len_pos, tag_len_pos, der);
 }
 
+DECLARE_ENCODE_FUNCTION(SubjectKeyIdentifier) {
+  Encode(val.key_identifier(), der);
+}
+
+DECLARE_ENCODE_FUNCTION(AuthorityKeyIdentifier) {
+  if (val.has_key_identifier()) {
+    size_t pos_of_tag = der.size();
+    Encode(val.key_identifier(), der);
+    // |key_identifier| is Context-specific with tag number 0 (RFC
+    // 5280, 4.2.1.1).
+    ReplaceTag(kAsn1ContextSpecific | 0x00, pos_of_tag, der);
+  }
+  if (val.has_authority_cert_issuer()) {
+    size_t pos_of_tag = der.size();
+    Encode(val.authority_cert_issuer(), der);
+    // |authority_cert_issuer| is Context-specific with tag number 1 (RFC
+    // 5280, 4.2.1.1).
+    ReplaceTag(kAsn1ContextSpecific | 0x01, pos_of_tag, der);
+  }
+  if (val.has_authority_cert_serial_number()) {
+    size_t pos_of_tag = der.size();
+    Encode(val.authority_cert_serial_number(), der);
+    // |authority_cert_serial_number| is Context-specific with tag number 2 (RFC
+    // 5280, 4.2.1.1).
+    ReplaceTag(kAsn1ContextSpecific | 0x02, pos_of_tag, der);
+  }
+}
+
+DECLARE_ENCODE_FUNCTION(RawExtension) {
+  if (val.has_pdu()) {
+    Encode(val.pdu(), der);
+    EncodeTagAndLength(kAsn1OctetString, der.size() - tag_len_pos, tag_len_pos,
+                       der);
+  } else {
+    Encode(val.extn_value(), der);
+  }
+}
+
+void EncodeExtensionValue(const Extension& val, std::vector<uint8_t>& der) {
+  switch (val.types_case()) {
+    case Extension::TypesCase::kAuthorityKeyIdentifier:
+      Encode(val.authority_key_identifier(), der);
+      break;
+    case Extension::TypesCase::kSubjectKeyIdentifier:
+      Encode(val.subject_key_identifier(), der);
+      break;
+    case Extension::TypesCase::TYPES_NOT_SET:
+      Encode(val.raw_extension(), der);
+      break;
+  }
+}
+
+void EncodeExtensionID(const Extension& val, std::vector<uint8_t>& der) {
+  if (val.has_extn_id()) {
+    Encode(val.extn_id(), der);
+    return;
+  }
+
+  std::vector<uint8_t> encoded_oid;
+  switch (val.types_case()) {
+    case Extension::TypesCase::kAuthorityKeyIdentifier:
+      // RFC 5280, 4.2.1.1: |AuthorityKeyIdentifier| OID is {2 5 29 35}.
+      encoded_oid = {(2 * 40) + 5, 29, 35};
+      break;
+    case Extension::TypesCase::kSubjectKeyIdentifier:
+      // RFC 5280, 4.2.1.2: |SubjectKeyIdentifier| OID is {2 5 29 14}.
+      encoded_oid = {(2 * 40) + 5, 29, 14};
+      break;
+    case Extension::TypesCase::TYPES_NOT_SET:
+      encoded_oid = {};
+      break;
```

Comment:
You could replace this with something clearer:

```
      Encode(val.raw_extension().extn_id(), der);
      return;
```

We can totally early return in the switch case. This lets you remove the conditional on 119 and just have line 120 be the end of this function, with no need for line 124.

That is: assume `encoded_oid` is always set by this switch case.

The current logic makes it seem like it’s valid to have an empty OID, but by the end of this switch, it’s a bug if the OID isn’t set.